contracts/vaults/balancer/internal/pool/TwoTokenPoolUtils.sol

```diff
@@ -121,37 +94,29 @@ library TwoTokenPoolUtils {
     /// @return primaryAmount primary token balance
     function _getTimeWeightedPrimaryBalance(
         TwoTokenPoolContext memory poolContext,
-        OracleContext memory oracleContext,
+        StableOracleContext memory oracleContext,
         StrategyContext memory strategyContext,
         uint256 bptAmount
     ) internal view returns (uint256 primaryAmount) {
-        // Gets the BPT token price denominated in token index = 0
-        uint256 bptPrice = BalancerUtils._getTimeWeightedOraclePrice(
-            address(poolContext.basePool.pool),
-            IPriceOracle.Variable.BPT_PRICE,
-            oracleContext.oracleWindowInSeconds
+        uint256 oraclePairPrice = _getOraclePairPrice(poolContext, strategyContext.tradingModule);
+        uint256 spotPrice = Stable2TokenOracleMath._getSpotPrice(
+            oracleContext, poolContext, poolContext.primaryIndex
         );
 
-        uint256 pairPrice = _getOraclePairPrice(poolContext, oracleContext, strategyContext.tradingModule);
-        uint256 primaryPrecision = 10 ** poolContext.primaryDecimals;
+        // Make sure spot price is within oracleDeviationLimit of pairPrice
+        Stable2TokenOracleMath._checkPriceLimit(strategyContext, oraclePairPrice, spotPrice);
+        
+        // Get shares of primary and secondary balances with the provided bptAmount
+        uint256 totalBPTSupply = poolContext.basePool.pool.totalSupply();
+        uint256 primaryBalance = poolContext.primaryBalance * bptAmount / totalBPTSupply;
+        uint256 secondaryBalance = poolContext.secondaryBalance * bptAmount / totalBPTSupply;
 
-        if (poolContext.primaryIndex == 0) {
-            // Since bptPrice is always denominated in the first token, we can just multiply by
-            // the amount in this case. Both bptPrice and bptAmount are in 1e18 but we need to scale
-            // this back to the primary token's native precision.
-            // underlyingValue = (bptPrice * bptAmount * primaryPrecision) / (1e18 * 1e18)
-            primaryAmount = (bptPrice * bptAmount * primaryPrecision) / 
-                BalancerConstants.BALANCER_PRECISION_SQUARED;
-        } else {
-            // The second token in the BPT pool is the price that we want to get. In this case, we need to
-            // convert secondaryTokenValue to underlyingValue using the pairPrice.
-            // Both bptPrice and bptAmount are in 1e18
-            uint256 secondaryAmount = (bptPrice * bptAmount) / BalancerConstants.BALANCER_PRECISION;
-
-            // And then normalizing to primary token precision we add:
-            // PrimaryAmount = (SecondaryAmount * primaryPrecision) / PairPrice
-            primaryAmount = (secondaryAmount * primaryPrecision) / pairPrice;
-        }
+        // Value the secondary balance in terms of the primary token using the oraclePairPrice
+        uint256 secondaryAmountInPrimary = secondaryBalance * BalancerConstants.BALANCER_PRECISION / spotPrice;
```

Comment:
@weitianjie2000 please use the oraclePairPrice here, not the spotPrice

Reply:
fixed